src/basic/ether-addr-util.h

```diff
@@ -1,11 +1,25 @@
 /* SPDX-License-Identifier: LGPL-2.1+ */
 #pragma once
 
+#include <linux/if_infiniband.h>
 #include <net/ethernet.h>
 #include <stdbool.h>
 
 #include "hash-funcs.h"
 
+#define HW_ADDR_MAX_SIZE CONST_MAX(ETH_ALEN, INFINIBAND_ALEN)
+
+union hw_addr_union {
+        struct ether_addr ether;
+        uint8_t infiniband[INFINIBAND_ALEN];
+        uint8_t bytes[HW_ADDR_MAX_SIZE];
+};
+
+typedef struct hw_addr_data {
+        union hw_addr_union addr;
+        size_t length;
+} hw_addr_data;
```

Comment:
why this? shouldn't the size be clear from context?

Reply:
mayba add:

```c
static inline size_t hw_addr_size(int iftype) {
        if (iftype == ARPHRD_ETHER)
                return sizeof(struct ether_addr);
        if (iftype == ARPHRD_INFINIBAND)
                return INFINIBAND_ALEN;
        return 0;
}
```

or so, to have a simple macro-like helper that derives the right size from the iftype?
}

Reply:
I think it's worth having the size there still, even if purely as a sanity check.
There's a couple places that sanity-check the address length against the interface type already, so if that length then gets derived from the iftype initially, that becomes kinda pointless.

It also simplifies the code a bit, specially when the hw_addr is to be used in places that right now use the iftype.

Reply:
Hmm, I agree with the suggestion by @poettering. The functions `netlink_message_read_in_addr_union()` or `_append_` take family `AF_INET` or `AF_INET6`. So, similarly, the new functions should take iftype `ARPHRD_ETHER` or `ARPHRD_INFINIBAND`.

Reply:
I'm really not sure about that.
Doing it that way would add a lot of complexity and logic in places that don't otherwise need it.
See for example the print functions I just added and their example usage in the networkctl status output.
That code right now does not need to care about the iftype, but if it would, it would need changed every time a new iftype gets supported.

Reply:
I'm also somewhat afraid that changing it to behave like it does for in_addr_union, where the family, or in this case iftype, is passed along everywhere causes a chain of regressions where formally implicitly supported interface types that behave like ethernet (I see references to CAN, PPP, WiFi-Devices and probably some tunnel devices) will then cause assertions because it's neither ETHER nor INFINIBAND, and compiling the full list everywhere seems way worse than just carrying along the size.

Reply:
OK. After grepping kernel codes, the length of `IFLA_ADDRESS` may have variety of length. I changed my mind. Let's keep this as is at least now.